addOns/graphql/src/main/java/org/zaproxy/addon/graphql/GraphQlParser.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2020 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.addon.graphql;
+
+import java.awt.EventQueue;
+import org.apache.log4j.Logger;
+import org.parosproxy.paros.control.Control;
+import org.parosproxy.paros.extension.history.ExtensionHistory;
+import org.parosproxy.paros.model.HistoryReference;
+import org.parosproxy.paros.model.Model;
+import org.parosproxy.paros.network.HttpMessage;
+
+public class GraphQlParser {
+
+    private static final Logger LOG = Logger.getLogger(GraphQlParser.class);
+
+    public static void persistMessage(final HttpMessage message) {
+        // Add the message to the history panel and sites tree
+        final HistoryReference historyRef;
+
+        try {
+            historyRef =
+                    new HistoryReference(
+                            Model.getSingleton().getSession(),
+                            HistoryReference.TYPE_ZAP_USER,
+                            message);
+        } catch (Exception e) {
+            LOG.error(e.getMessage(), e);
+            return;
+        }
+
+        final ExtensionHistory extHistory =
+                Control.getSingleton().getExtensionLoader().getExtension(ExtensionHistory.class);
+        if (extHistory != null) {
+            EventQueue.invokeLater(
+                    new Runnable() {
+
+                        @Override
+                        public void run() {
+                            extHistory.addHistory(historyRef);
+                            Model.getSingleton()
+                                    .getSession()
+                                    .getSiteTree()
+                                    .addPath(historyRef, message);
```

Comment:
addPath needs to be done in the Event Dispatch Thread (this PR might help https://github.com/secdec/attack-surface-detector-zap/pull/14/). addHistory doesn't need to be done in EDT but can't hurt.

This class will probably be used by the API so it might be worth wrapping the site tree addition in a view check.

Reply:
There's a core method `ThreadUtils#invokeAndWait` that handles that (it's synchronous but this methods should be as well?).

Reply:
I used invokeAndWait to do addPath in the EDT, but getting this error: Exception in thread "AWT-EventQueue-0"java.lang.Error: Cannot call invokeAndWait from the event dispatcher thread. It seems that it is already in the EDT, so there is no need to use invokeAndWait... What should be done instead?